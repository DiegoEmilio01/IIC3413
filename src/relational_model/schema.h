#pragma once

#include <algorithm>
#include <cassert>
#include <string>
#include <vector>

// IIC3413-DB will only support as datatypes:
// - strings up to 255 bytes length in UTF-8 representation
// - 64 bit signed integer
enum class DataType {
    STR,
    INT
};


class Schema {
public:
    Schema(std::vector<std::string>&& column_names,
           std::vector<DataType>&&    datatypes) :
        column_names (std::move(column_names)),
        datatypes    (std::move(datatypes))
    {
        assert(column_names.size() == datatypes.size());
    }

    Schema() {}

    std::vector<std::string> column_names;

    std::vector<DataType> datatypes;

    bool operator==(const Schema& other) const {
        return datatypes == other.datatypes && column_names == other.column_names;
    }
};
