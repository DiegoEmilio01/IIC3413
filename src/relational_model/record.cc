#include "record.h"

#include <cassert>
#include <cstring>

Record::Record(std::vector<DataType>&& _types) :
    types (std::move(_types))
{
    for (auto type : types) {
        switch (type) {
        case DataType::INT: {
            values.push_back(OutValue((int64_t) 0));
            break;
        }
        case DataType::STR: {
            char* str_buffer = new char[MAX_STRLEN+1];
            std::memset(str_buffer, '\0', MAX_STRLEN+1);
            values.push_back(OutValue(str_buffer));
            break;
        }
        }
    }
}


Record::~Record() {
    auto index = 0;
    for (auto type : types) {
        switch (type) {
        case DataType::INT: {
            // no need to free
            break;
        }
        case DataType::STR: {
            delete[] values[index].str_value;
            break;
        }
        }
        index++;
    }
}

void Record::set(const std::vector<std::variant<std::string, int64_t>>& new_values) {
    assert(new_values.size() == values.size());
    assert(new_values.size() == types.size());

    for (size_t i = 0; i < values.size(); i++) {
        auto& value = new_values[i];

        switch (types[i]) {
        case DataType::INT: {
            assert(std::holds_alternative<int64_t>(value));
            values[i].int_value = std::get<int64_t>(value);
            break;
        }
        case DataType::STR: {
            assert(std::holds_alternative<std::string>(value));
            auto& str = std::get<std::string>(value);
            assert(str.size() <= Record::MAX_STRLEN);
            memcpy(values[i].str_value, str.c_str(), str.size()+1);
            break;
        }
        }
    }
}
