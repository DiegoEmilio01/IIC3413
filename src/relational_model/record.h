#pragma once

#include <cassert>
#include <cstdint>
#include <ostream>
#include <variant>
#include <vector>

#include "relational_model/schema.h"

union OutValue {
    char* str_value;
    int64_t int_value;

    explicit OutValue(char* str_value) : str_value(str_value) {}
    explicit OutValue(int64_t int_value) : int_value(int_value) {}
};

class Record {
public:
    static constexpr int64_t MAX_STRLEN = 255;

    Record(std::vector<DataType>&& types);

    ~Record();

    void set(const std::vector<std::variant<std::string, int64_t>>& values);

    friend std::ostream& operator<<(std::ostream& os, const Record& o) {
        assert(o.types.size() == o.values.size());

        char separator[2] = {'\0', '\0'};
        for (unsigned i = 0; i < o.types.size(); i++) {
            os << separator;
            switch (o.types[i]) {
            case DataType::INT: {
                os << o.values[i].int_value;
                break;
            }
            case DataType::STR:
                os << '"' << o.values[i].str_value << '"';
                break;
            }
            separator[0] = ',';
        }
        return os;
    }

    std::vector<OutValue> values;

    std::vector<DataType> types;
};
