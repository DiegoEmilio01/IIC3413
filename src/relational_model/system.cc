#include "system.h"

#include <stdexcept>

// memory for the global objects
static typename std::aligned_storage<sizeof(BufferManager), alignof(BufferManager)>::type buffer_mgr_buf;
static typename std::aligned_storage<sizeof(FileManager), alignof(FileManager)>::type     file_mgr_buf;
static typename std::aligned_storage<sizeof(Catalog), alignof(Catalog)>::type             catalog_buf;

BufferManager& buffer_mgr  = reinterpret_cast<BufferManager&>(buffer_mgr_buf);
FileManager& file_mgr      = reinterpret_cast<FileManager&>(file_mgr_buf);
Catalog& catalog           = reinterpret_cast<Catalog&>(catalog_buf);


System::System(const std::string& db_folder, uint64_t buffer_size) {
    new (&file_mgr) FileManager(db_folder);
    new (&buffer_mgr) BufferManager(buffer_size);
    new (&catalog) Catalog("catalog.dat");
}

System System::init(const std::string& db_folder, uint64_t buffer_size) {
    if (initialized) {
        throw std::logic_error("System::init cannot be called twice");
    }
    initialized = true;
    return System(db_folder, buffer_size);
}

System::~System() {
    catalog.~Catalog();
    buffer_mgr.~BufferManager();
    file_mgr.~FileManager();
}