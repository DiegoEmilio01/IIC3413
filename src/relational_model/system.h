#pragma once

#include "relational_model/catalog.h"
#include "storage/buffer_manager.h"
#include "storage/file_manager.h"

// class SystemScope {
// public:
//     ~SystemScope();
// };

class System {
public:
    ~System();

    // static SystemScope init();
    static System init(const std::string& db_folder, uint64_t buffer_size);

private:
    static inline bool initialized = false;

    System(const std::string& db_folder, uint64_t buffer_size);
};


// global objects
extern FileManager& file_mgr;
extern BufferManager& buffer_mgr;
extern Catalog& catalog;
