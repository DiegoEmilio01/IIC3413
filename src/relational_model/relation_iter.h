#pragma once

#include "relational_model/record.h"

class RelationIter {
public:
    virtual ~RelationIter() = default;

    virtual void begin(Record& out) = 0;

    virtual bool next() = 0;

    virtual void reset() = 0;
};
