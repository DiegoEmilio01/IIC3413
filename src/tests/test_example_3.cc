#include <iostream>

#include "relational_model/record.h"
#include "relational_model/relation_iter.h"
#include "relational_model/schema.h"
#include "relational_model/system.h"

constexpr uint64_t GB = 1024 * 1024 * 1024; // 1 GB

int main() {
    uint64_t buffer_size = 1 * GB;
    std::string database_folder = "data/test_example";

    // Need to call System::init before start using the database
    // When this object comes out of scope the database is no longer usable
    auto system = System::init(database_folder, buffer_size);

    Schema table1_schema(
        { "col1", "col2" },
        { DataType::STR, DataType::INT }
    );

    Schema existing_table_schema;
    HeapFile* table = catalog.get_table("test_1", &existing_table_schema);

    if (table == nullptr) { // table doesn't exist
        table = catalog.create_table("test_1", table1_schema);
    } else {
        assert(existing_table_schema == table1_schema);
    }

    Record record_buf({DataType::STR, DataType::INT});

    record_buf.set({"test_record_1", 1});
    auto r1 = table->insert_record(record_buf);

    record_buf.set({"test_record_2", 2});
    table->insert_record(record_buf);

    record_buf.set({"test_record_3", 3});
    auto r3 = table->insert_record(record_buf);

    record_buf.set({"test_record_4", 4});
    table->insert_record(record_buf);

    table->delete_record(r3);

    record_buf.set({"test_record_5", 5});
    table->insert_record(record_buf);

    table->delete_record(r1);

    record_buf.set({"test_record_6", 6});
    table->insert_record(record_buf);

    auto table_iter = table->get_record_iter();

    table_iter->begin(record_buf);

    while (table_iter->next()) {
        // print result
        std::cout << record_buf << "\n";
    }

    return 0;
}
