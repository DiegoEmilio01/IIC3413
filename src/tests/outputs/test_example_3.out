"test_record_6",6
"test_record_2",2
"test_record_5",5
"test_record_4",4
